### PR TITLE
x/twap: three asset pool tests (part 1 of 2)

### PR DESCRIPTION
**Pull request by czarcas7ic**
Closes: #2538

## What is the purpose of the change

PR #2604 attempted to combine current tests with three asset pool tests. This proved to be very messy and unclear. This PR is another attempt at implementing these tests in a separate and clearer manner. Additionally, more complex test cases are used (varying spot prices for all three assets) for every three asset pool test case along with comments explaining the math behind each expected twap.

Part one addresses api_test.go

Part two will address logic_test.go and store_test.go

## Brief Changelog

- Adds test cases for api_test.go
- Adds function in keeper_test.go to generate default three asset pools

## Testing and Verifying

This change is already covered by existing tests

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? not applicable

### CODE REVIEW

File: x/twap/api_test.go [hunk 1]
```diff
@@ -11,26 +11,43 @@ import (
...
 	ThreePlusOneThird sdk.Dec = sdk.MustNewDecFromStr("3.333333333333333333")
 
 	// base record is a record with t=baseTime, sp0=10(sp1=0.1) accumulators set to 0
 	baseRecord types.TwapRecord = newTwapRecordWithDefaults(baseTime, sdk.NewDec(10), sdk.ZeroDec(), sdk.ZeroDec())
 
-	// accum0 = 10 seconds * (spot price = 10) = OneSec * 10 * 10
-	// accum1 = 10 seconds * (spot price = 0.1) = OneSec
-	accum0, accum1 sdk.Dec = OneSec.MulInt64(10 * 10), OneSec
+	tapRecordAB, tapRecordAC, tapRecordBC = newThreeAssetPoolTwapRecordWithDefaults(
+		baseTime, sdk.NewDec(10), sdk.ZeroDec(), sdk.ZeroDec(), sdk.ZeroDec())
```
Comment (p0mvn) [inline x/twap/api_test.go:27]: nit:
```suggestion
	tapRecordAB, tapRecordAC, tapRecordBC = newThreeAssetPoolTwapRecordWithDefaults(
		baseTime,
		sdk.NewDec(10), // spot price 0
		sdk.ZeroDec(),  // accum A
		sdk.ZeroDec(),  // accum B
		sdk.ZeroDec())  // accum C
```

File: x/twap/api_test.go [hunk 2]
```diff
@@ -11,26 +11,43 @@ import (
...
+
+	tPlus10sp5TapRecordAB, tPlus10sp5TapRecordAC, tPlus10sp5TapRecordBC = newThreeAssetPoolTwapRecordWithDefaults(
+		baseTime.Add(10*time.Second), sdk.NewDec(5), accumA, accumB, accumC)
 
 	// accumulators updated from tPlus10sp5Record with
 	// t = baseTime + 20
-	// sp0 = 2, sp1 = 0.5
+	// spA = 2, spB = 0.5, spC = 4
 	tPlus20sp2Record = newTwapRecordWithDefaults(
 		baseTime.Add(20*time.Second), sdk.NewDec(2), OneSec.MulInt64(10*10+5*10), OneSec.MulInt64(3))
```
Comment (p0mvn) [inline x/twap/api_test.go:47]: nit: to reduce cognitive overhead
```suggestion
	tPlus20sp2Record = newTwapRecordWithDefaults(
		baseTime.Add(20*time.Second),
		sdk.NewDec(2),               // spot price 0
		OneSec.MulInt64(10*10+5*10), // accum A
		OneSec.MulInt64(3))          // accum B
```

File: x/twap/api_test.go [hunk 3]
```diff
@@ -11,26 +11,43 @@ import (
...
+
+	tPlus20sp2TapRecordAB, tPlus20sp2TapRecordAC, tPlus20sp2TapRecordBC = newThreeAssetPoolTwapRecordWithDefaults(
+		baseTime.Add(20*time.Second), sdk.NewDec(2), OneSec.MulInt64(10*10+5*10), OneSec.MulInt64(3), OneSec.MulInt64(20*10+10*10))
```
Comment (p0mvn) [inline x/twap/api_test.go:50]:
```suggestion
	tPlus20sp2TapRecordAB, tPlus20sp2TapRecordAC, tPlus20sp2TapRecordBC = newThreeAssetPoolTwapRecordWithDefaults(
		baseTime.Add(20*time.Second),
		sdk.NewDec(2),                // spot price 0
		OneSec.MulInt64(10*10+5*10),  // accum A
		OneSec.MulInt64(3),           // accum B
		OneSec.MulInt64(20*10+10*10)) // accum C
```

File: x/twap/api_test.go [hunk 4]
```diff
@@ -11,26 +11,43 @@ import (
...
+		baseTime.Add(10*time.Second), sdk.NewDec(5), accumA, accumB)
+
+	tPlus10sp5TapRecordAB, tPlus10sp5TapRecordAC, tPlus10sp5TapRecordBC = newThreeAssetPoolTwapRecordWithDefaults(
+		baseTime.Add(10*time.Second), sdk.NewDec(5), accumA, accumB, accumC)
 
 	// accumulators updated from tPlus10sp5Record with
 	// t = baseTime + 20
-	// sp0 = 2, sp1 = 0.5
+	// spA = 2, spB = 0.5, spC = 4
 	tPlus20sp2Record = newTwapRecordWithDefaults(
```
Comment (p0mvn) [inline x/twap/api_test.go:46]:
```suggestion
	tPlus20sp2Record = newTwoAssetPoolTwapRecordWithDefaults(
```

File: x/twap/api_test.go [hunk 5]
```diff
@@ -103,13 +120,30 @@ func makeSimpleTwapInput(startTime time.Time, endTime time.Time, isQuoteTokenA b
 	return getTwapInput{1, quoteAssetDenom, baseAssetDenom, startTime, endTime}
 }
 
+func makeSimpleTapTwapInput(startTime time.Time, endTime time.Time, baseQuoteAB, baseQuoteCA, baseQuoteBC bool) []getTwapInput {
```
Comment (p0mvn) [inline x/twap/api_test.go:123]: What does "Tap" mean? I'm guessing something to do with 3-assets?

Please consider being more specific with the abbreviation

File: x/twap/api_test.go [hunk 6]
```diff
@@ -103,13 +120,30 @@ func makeSimpleTwapInput(startTime time.Time, endTime time.Time, isQuoteTokenA b
...
+	var twapInput []getTwapInput
+	quoteAssetDenom, baseAssetDenom := denom0, denom1
+	if baseQuoteAB {
+		baseAssetDenom, quoteAssetDenom = quoteAssetDenom, baseAssetDenom
+	}
```
Comment (p0mvn) [inline x/twap/api_test.go:128]: Consider encapsulating into a function to reduce code duplication

File: x/twap/keeper_test.go [hunk 7]
```diff
@@ -373,6 +373,33 @@ func newTwapRecordWithDefaults(t time.Time, sp0, accum0, accum1 sdk.Dec) types.T
 	}
 }
 
+func newThreeAssetPoolTwapRecordWithDefaults(t time.Time, sp0, accumA, accumB, accumC sdk.Dec) (types.TwapRecord, types.TwapRecord, types.TwapRecord) {
```
Comment (p0mvn) [inline x/twap/keeper_test.go:376]: Please add godoc specifying how the rest of the spot prices are inferred from `sp0`

File: x/twap/api_test.go [hunk 8]
```diff
@@ -254,6 +288,71 @@ func (s *TestSuite) TestGetArithmeticTwap() {
...
+		"(3 record) interpolate: get twap closer to second record": {
+			recordsToSet: []types.TwapRecord{tapRecordAB, tapRecordAC, tapRecordBC,
+				tPlus10sp5TapRecordAB, tPlus10sp5TapRecordAC, tPlus10sp5TapRecordBC,
+				tPlus20sp2TapRecordAB, tPlus20sp2TapRecordAC, tPlus20sp2TapRecordBC},
+			ctxTime: tPlusOneMin,
+			input:   makeSimpleTapTwapInput(baseTime.Add(15*time.Second), baseTime.Add(30*time.Second), baseQuoteAB, baseQuoteCA, baseQuoteBC),
+			// A 5 for 5s, 2 for 10s = 45/15 = 3
+			// C 10 for 5s, 4 for 10s = 140/15 = 6
+			// B .2 for 5s, .5 for 10s = 7/15 = .4
+			expTwap: []sdk.Dec{sdk.NewDec(3), sdk.NewDec(6), sdk.NewDecWithPrec(4, 1)}, // 5 for 5s, 2 for 10s = 45 / 15 = 3
```
Comment (p0mvn) [inline x/twap/api_test.go:330]: Incredibly clear, nice work!

Comment (p0mvn) [inline x/twap/api_test.go:330]: nit: I think that a copy of this is already present above
```suggestion
			expTwap: []sdk.Dec{sdk.NewDec(3), sdk.NewDec(6), sdk.NewDecWithPrec(4, 1)},
```

**Review by p0mvn — APPROVED** [review]

File: x/twap/api_test.go (hunk 5 above)
Comment (czarcas7ic) [reply x/twap/api_test.go:134]: "ThreeAssetPool", do you suggest I just error on the side of verbosity and change every reference of "tap" to "threeAssetPool"? The more I think about it the more I feel I should change it to reduce cognitive overhead

Comment (czarcas7ic) [reply x/twap/api_test.go:134]: I decided to change `tap` to threeAsset, hopefully that is a good middle ground!

**Comment by czarcas7ic** [comment]
If there are no further comments, I will merge this by end of day to unblock the merging of part 2 🙂

**Pull request merged**